source/common/network/dns_impl.cc

```diff
@@ -216,13 +259,230 @@ ActiveDnsQuery* DnsResolverImpl::resolve(const std::string& dns_name,
+                                                unsigned char* abuf, int alen) {
+  auto hquery = static_cast<HostQuery*>(arg);
+  hostent* host = nullptr;
+  void* addrttls = nullptr;
+  int naddrttls = addrttl_len_;
+
+  hquery->timeouts_ += timeouts;
+  if (status == ARES_SUCCESS) {
+    if (hquery->sent_family_ == AF_INET) {
+      addrttls = new ares_addrttl[naddrttls];
+      status =
+          ares_parse_a_reply(abuf, alen, &host, static_cast<ares_addrttl*>(addrttls), &naddrttls);
+    } else if (hquery->sent_family_ == AF_INET6) {
+      addrttls = new ares_addr6ttl[naddrttls];
+      status = ares_parse_aaaa_reply(abuf, alen, &host, static_cast<ares_addr6ttl*>(addrttls),
+                                     &naddrttls);
+      if ((status == ARES_ENODATA || status == ARES_EBADRESP ||
+           (status == ARES_SUCCESS && host && host->h_addr_list[0] == nullptr)) &&
+          hquery->want_family_ == AF_UNSPEC) {
+        // The query returned something but either there were no AAAA records (e.g. just CNAME) or
+        // the response was malformed. Try looking up A instead.
+        if (host) {
+          ares_free_hostent(host);
+        }
+        delete[] static_cast<ares_addr6ttl*>(addrttls);
+
+        hquery->sent_family_ = AF_INET;
+        ares_search(hquery->channel_, hquery->name_.c_str(), C_IN, T_A, hostCallback, hquery);
+        return;
+      }
+    }
+    endHquery(hquery, status, host, addrttls, naddrttls);
+  } else if ((status == ARES_ENODATA || status == ARES_EBADRESP || status == ARES_ETIMEOUT) &&
+             (hquery->sent_family_ == AF_INET6 && hquery->want_family_ == AF_UNSPEC)) {
+    // The AAAA query yielded no useful result. Now look up an A instead.
+    hquery->sent_family_ = AF_INET;
+    ares_search(hquery->channel_, hquery->name_.c_str(), C_IN, T_A, hostCallback, hquery);
+  } else if (status == ARES_EDESTRUCTION) {
+    endHquery(hquery, status, nullptr, nullptr, 0);
+  } else {
+    nextLookup(hquery, status);
+  }
+}
+
+void DnsResolverImpl::AresWrapper::endHquery(HostQuery* hquery, int status, hostent* host,
+                                             void* addrttls, int naddrttls) {
+  hquery->callback_(hquery->arg_, status, hquery->timeouts_, host, addrttls, naddrttls);
+  if (host != nullptr) {
+    ares_free_hostent(host);
+  }
+
+  if (addrttls != nullptr) {
+    if (hquery->sent_family_ == AF_INET) {
+      delete[] static_cast<ares_addrttl*>(addrttls);
+    } else {
+      delete[] static_cast<ares_addr6ttl*>(addrttls);
+    }
+  }
+
+  delete hquery;
+}
+
+bool DnsResolverImpl::AresWrapper::fakeHostent(const std::string& name, int family,
+                                               AresHostCallback callback, void* arg) {
+  struct hostent hostent;
+  char* aliases[1] = {nullptr};
+  char* addrs[2];
+  int result = 0;
+  struct in_addr in;
+  struct ares_in6_addr in6;
+
+  if (family == AF_INET || family == AF_INET6) {
```

Comment:
Is this code introduced by your PR or you can find it somewhere else?

Reply:
No. It's in c-ares library.
https://github.com/c-ares/c-ares/blob/9f1fdbf5dd633f81352fac0d6bc0d0c4d45be459/ares_gethostbyname.c#L244